scripts/build_test_history.py

```diff
@@ -0,0 +1,447 @@
+        # The process start date entity is only updated once all aligned runs
+        # for a given revision are processed.
+        if revision not in revisions_processed:
+            revisions_processed[revision] = {
+                'chrome': False,
+                'edge': False,
+                'firefox': False,
+                'safari': False,
+            }
+
+        if should_process_run(run_metadata):
+            process_single_run(run_metadata)
+        else:
+            print('Run has already been processed')
+
+        revisions_processed[revision][browser_name] = True
+        # If all runs for this revision have been processed, we can update
+        # the most recently processed date to the run's start time.
+        if (revisions_processed[revision]['chrome'] and
+                revisions_processed[revision]['edge'] and
+                revisions_processed[revision]['firefox'] and
+                revisions_processed[revision]['safari']):
+            print(f'All browsers have been processed for {revision}. '
+                  'Updating date.')
+            update_recent_processed_date(
+                process_start_entity, run_metadata['time_start'])
+    print('Set of runs processed after '
+          f'{round(time.time() - start, 0)} seconds.')
+
+
+# Get the list of metadata for the most recent aligned runs.
+def get_aligned_run_info(
+        date_entity: MostRecentHistoryProcessed) -> Optional[list]:
+    date_start = date_entity.Date
+    date_start_obj = datetime.strptime(date_start, '%Y-%m-%dT%H:%M:%S.%fZ')
+
+    # Since aligned runs need to all be completed runs to be fetched,
+    # a time window buffer of 24 hours is kept to allow runs to finish before
+    # assuming we've processed all aligned runs up to present time.
+    # Therefore, we only process runs up to (now - 24 hours).
+    yesterday = datetime.now() - timedelta(days=1)
+    end_interval = date_start_obj + timedelta(days=1)
+    if end_interval > yesterday:
+        end_interval = yesterday
+
+    end_interval_string = end_interval.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
+    url = (f'{RUNS_API_URL}?label=master'
+           '&label=experimental&max-count=1&aligned'
+           f'&from={date_start}&to={end_interval_string}')
+
+    try:
+        resp = requests.get(url)
+    # Sometimes this request can time out. If it does, just return
+    # an empty list and attempt the fetch again.
+    except requests.exceptions.ReadTimeout:
+        return []
+    runs_list: list[MetadataDict] = resp.json()
+
+    # If we have no runs to process in this date interval,
+    # we can skip this interval for processing from now on.
+    if len(runs_list) == 0:
+        print('No runs found for this interval.')
+        update_recent_processed_date(date_entity, end_interval_string)
+        # If we've processed up to (now - 24 hours), then return null,
+        # which signals we're done processing.
+        if end_interval == yesterday:
```

Comment:
Shouldn't we return here regardless of `end_interval == yesterday condition`? runs_list has nothing to sort in line 360

Reply:
True! I added another statement to make sure we return here regardless